src/main/scala/shine/DPIA/showDPIA.scala

```diff
@@ -0,0 +1,86 @@
+package shine.DPIA
+
+import shine.DPIA.Phrases._
+import shine.DPIA.Types._
+import shine.DPIA.primitives.imperative._
+import shine.DPIA.primitives.functional._
+
+sealed trait ShowTypes
+case object On extends ShowTypes
+case object Off extends ShowTypes
+
+object showDPIA {
+  def apply[T <: PhraseType](p : Phrase[T]) : String = showDPIA().showPhrase(p)
+  def apply[T <: PhraseType](p : Phrase[T], showTypes : ShowTypes) : String = showDPIA(showTypes).showPhrase(p)
+}
+
+case class showDPIA(showTypes : ShowTypes = Off) {
+  val TAB : String = "    "
+  def tabulate : String => String =
+    _.linesIterator.map(l => if (l.isEmpty) l else TAB + l).mkString("\n")
+
+  def showTypeId : Kind.Identifier => String = _.name
+  def showKind : Kind[_, _] => String = _.name
+  def showUnary : Operators.Unary.Value => String = _.toString
+  def showBinary : Operators.Binary.Value => String = _.toString
+  def showData : Data => String = _.toString
+  def showNat : Nat => String = _.toString
+  def showPhraseType : PhraseType => String = _.toString
+  def showDataType : DataType => String = _.toString
+  def showAddressSpace : AddressSpace => String = _.toString
+  def showAccess : AccessType => String = _.toString
+  def showNatToNat : NatToNat => String = _.toString
+  def showNatToData : NatToData => String = _.toString
+  def showTypeDo[T](kind : Kind[T, _], arg : T) : String = kind match {
+    case PhraseKind => showPhraseType(arg.asInstanceOf[PhraseType])
+    case DataKind => showDataType(arg.asInstanceOf[DataType])
+    case NatKind => showNat(arg.asInstanceOf[Nat])
+    case AddressSpaceKind => showAddressSpace(arg.asInstanceOf[AddressSpace])
+    case AccessKind => showAccess(arg.asInstanceOf[AccessType])
+    case NatToNatKind => showNatToNat(arg.asInstanceOf[NatToNat])
+    case NatToDataKind => showNatToData(arg.asInstanceOf[NatToData])
+  }
+  def showPrim : Primitive[_] => String = {
+    case Comment(s) => s"/* ${s} */"
+    case Seq(c1, c2) => s"{\n${tabulate(showPhrase(c1))};\n${tabulate(showPhrase(c2))}\n}"
+    case New(dt, f) => s"new (exp[${showDataType(dt)}] x acc[${showDataType(dt)}]) in ${showPhrase(f)}"
+    case Assign(dt, lhs, rhs) => s"${showTypedPhrase(DataKind, dt, lhs)} = ${showPhrase(rhs)}"
+    case IdxAcc(n, dt, idx, array) => s"${showTypedPhrase(DataKind, ArrayType(n, dt), array)}[${showPhrase(idx)}]"
+    case Idx(n, dt, idx, array) => s"${showTypedPhrase(DataKind, ArrayType(n, dt), array)}[${showPhrase(idx)}]"
+    case NatAsIndex(n, e) => s"${showTypedPhrase(NatKind, n, e)}"
+    case Drop(n, m, dt, e) => s"drop ${showNat(n)} ${showTypedPhrase(DataKind, ArrayType(n + m, dt), e)}"
+    case DropAcc(n, m, dt, e) => s"drop ${showNat(n)} ${showTypedPhrase(DataKind, ArrayType(n + m, dt), e)}"
+    case Take(n, m, dt, e) => s"take ${showNat(n)} ${showTypedPhrase(DataKind, ArrayType(n + m, dt), e)}"
```

Comment:
I think that this will create parenthesing ambiguities, maybe it should be printed as `(take ...)`?.

Reply:
done, for consistency I changed all scoping from `binder. { scope }` to `(binder. scope)`